### PR TITLE
[service.languagepreferencemanager] 1.0.1

### PR DESCRIPTION
**Pull request by rockrider69**
### A manager for audio and subtitle preferences

It provides an easy way to set your preferred audio streams and subtitle languages, usefull for Movies or TV-Shows which includes multiple such tracks.

You can select which audio tracks and subtitles to automatically activate based on your priorities, and define simple conditional rules like "if audio is xxx then activate subtitles yyy" via drop/down lists. Conditions can handle "forced" tracks tag.

More advanced custom rules can be defined as well like fre:non>eng:eng>any:fre>any:eng>.... to handle more than 3 priority levels (See changelog for more on the syntax).

Some new features added recently : reassess the rules on the fly whenever you switch audio track while watching, skip subtitles tracks base on a keyword list vs name content (ex; to skip Signs and Songs tracks if misordered, ...).

Note : this addon has been developed and used for several years. I took over the maintenance few months ago. See README for some credits to previous devs.
### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [script.foo.bar] 1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

**Comment by rockrider69** [comment]
@romanvm
Hello !
 I see that you're reviewing this new addon, just let me know if there is anything more I must do at this stage
regards

**Comment by rockrider69** [comment]
Hello !
I wouldn't want to spam, but I am starting to wonder if I misunderstood the status :  am I the one, the developer,  in "Waiting for developer to respond", to provide more information about the addon ?
Sorry if I did miss any point here !
Regards

**Comment by romanvm** [comment]
Sorry for the late reply but I see that the issue of the big addon entrypoint script has not been addressed. Please read my respective comment above.

**Comment by rockrider69** [comment]
No problem, thank you for taking the time, I realize this is a pretty big addon to review.
I do not master github very well yet because I can't find any previous comment from you or any issue about the "issue of the big addon entrypoint script" you refer to.
So far on this PR I see only 3 comments total :  2 from me and your most recent one from yesterday. And all checks are green.

Could you please guide me where I find your initial comment/request so I can try to understand the expectation and hopefully work on it?
Thanks in advance !

**Comment by romanvm** [comment]
@rockrider69  All comments should be visible above if you are logged in. In any case the problem is the big addon entrypoint script that may result in increased addon loading time on slower systems like Raspberry PI. A best practice is to have all logic in importable modules and keep the entrypoint as small as possible.

**Comment by rockrider69** [comment]
**@romanvm** not sure why but I am logged in but still can't see your very first comment...
Anyway thanks to your last comment and the help of the example I think I see what I have to do :)
Will start to work on it now and push an update later.

**Comment by rockrider69** [comment]
@romanvm
hello, I have restructured the code to have the smallest entrypoint as your recommanded, by moving all logic into a separate module.
I hope it is at least going into the right direction.
Let me know! 

still learning the git stuff as well, but I think I eventually managed to merge things into one single commit;.

File: service.languagepreferencemanager/resources/lib/prefutils.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+                                            ["currentaudiostream", "audiostreams", "subtitleenabled",
+                                             "currentsubtitle", "subtitles" ],
+                                            "playerid": activePlayerID },
+                                "id": 1}
+        details_query_string = simplejson.dumps(details_query_dict)
+        json_query = xbmc.executeJSONRPC(details_query_string)
+        #json_query = unicode(json_query, 'utf-8', errors='ignore')
+        json_response = simplejson.loads(json_query)
+        
+        if 'result' in json_response and json_response['result'] != None:
```
Comment (romanvm) [inline service.languagepreferencemanager/resources/lib/prefutils.py:337]: `json_response['result'] is not None`

File: service.languagepreferencemanager/resources/lib/prefutils.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+                                 "method": "Player.GetItem",
+                                 "params": { "properties":
+                                            ["genre", "tag"],
+                                            "playerid": activePlayerID },
+                                 "id": 1}
+        genre_tags_query_string = simplejson.dumps(genre_tags_query_dict)
+        json_query = xbmc.executeJSONRPC(genre_tags_query_string)
+        #json_query = unicode(json_query, 'utf-8', errors='ignore')
+        json_response = simplejson.loads(json_query)
+        if 'result' in json_response and json_response['result'] != None:
```
Comment (romanvm) [inline service.languagepreferencemanager/resources/lib/prefutils.py:360]: `is not None`

**Review by romanvm — APPROVED** [review]
LGTM, comments are not mandatory

**Comment by romanvm** [comment]
Sorry for the delay, now it looks OK.

**Comment by rockrider69** [comment]
> Sorry for the delay, now it looks OK.

No problem, above all thanks a lot for the guidance !
I'll keep track of your recommandations and integrate those changes into a next update.

File: service.languagepreferencemanager/resources/lib/langcodes.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (rockrider69) [reply service.languagepreferencemanager/resources/lib/langcodes.py:1]: Not mandatory hence will be fixed in next update.

File: service.languagepreferencemanager/resources/lib/prefutils.py (hunk 1 above)
Comment (rockrider69) [reply service.languagepreferencemanager/resources/lib/prefutils.py:337]: Not mandatory hence will be fixed in next update.

File: service.languagepreferencemanager/resources/lib/prefutils.py (hunk 2 above)
Comment (rockrider69) [reply service.languagepreferencemanager/resources/lib/prefutils.py:360]: Not mandatory hence will be fixed in next update.

**Pull request merged**